src/tests/profiler/gcheapenumeration/gcheapenumeration.cs

```diff
@@ -0,0 +1,144 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace Profiler.Tests
+{
+    class GCHeapEnumerationTests
+    {
+        static readonly string ShouldSetMonitorGCEventMaskEnvVar = "Set_Monitor_GC_Event_Mask";
+        static readonly Guid GCHeapEnumerationProfilerGuid = new Guid("8753F0E1-6D6D-4329-B8E1-334918869C15");
+        static List<object> _rootObjects = new List<object>();
+
+        [DllImport("Profiler")]
+        private static extern void EnumerateGCHeapObjects();
+
+        [DllImport("Profiler")]
+        private static extern void SuspendRuntime();
+
+        [DllImport("Profiler")]
+        private static extern void ResumeRuntime();
+
+        [DllImport("Profiler")]
+        private static extern void EnumerateHeapObjectsInBackgroundThread();
+
+        public static int EnumerateGCHeapObjectsSingleThreadNoPriorSuspension()
+        {
+            _rootObjects.Add(new CustomGCHeapObject());
+            EnumerateGCHeapObjects();
+            return 100;
+        }
+
+        public static int EnumerateGCHeapObjectsSingleThreadWithinProfilerRequestedRuntimeSuspension()
+        {
+            _rootObjects.Add(new CustomGCHeapObject());
+            SuspendRuntime();
```

Comment:
This doesn't look like it would work. If this SuspendRuntime() returns back to managed code that means the runtime isn't suspended. This seems like it either deadlocks or it isn't suspending.

I'd recommend making a native method that invokes all of these functions rather than trying to coordinate it from managed code.

Reply:
Its a surprising result. My best guess is that some optimization in how the JIT does p/invokes causes it to optimize away barrier that prevents returning to managed code in between adjacently called p/invokes. I would not recommend writing the test in a way that relies on that kind of optimization though. I'd still suggest writing a single native method that does Suspend(), Enumerate(), and Resume() clearly within native code.

Reply:
Added `EnumerateGCHeapObjectsWithinProfilerRequestedRuntimeSuspension` which does all three